### ISSUE TITLE
RFC: option for simp to create new goals

### ISSUE THREAD

**Issue by fpvandoorn** [leanprover/lean4#5994]
### Proposal

Implement a new option for `simp` that will cause `simp` to create new subgoals when a conditional simp-lemma is applied. This option is off by default, and is potentially dangerous unless called with `simp only`, but we might want it to be on by default in Mathlib's `simp_rw`.

This example shows the basic desired behavior:
```lean
example (p : Prop) (a b : Nat) (h : p → a = b) : a = b := by
  simp (config := {createSubgoals := true}) only [h] 
  -- desired output: 1 goal: ⊢ p
```
The behavior is of course more complicated after applying some `congr`-theorems. In that case, all the new variables in the local context should be available for the created subgoal. Example:
```lean
open Classical
noncomputable def sum (p : Nat → Prop) (f : Nat → Nat) : Nat := if p 0 then f 0 else 0

@[congr]
theorem sum_congr {p p' : Nat → Prop} {f f' : Nat → Nat}
    (hp : ∀ n, p n ↔ p' n) (hf : ∀ n, p' n → f n = f' n) :
    sum p f = sum p' f' := by
  simp (config := {contextual := true}) [sum, hp, hf]

example {p p' : Nat → Prop} {f f' : Nat → Nat} (hf : ∀ n, p' n → f n = f' n) :
    sum p (fun n ↦ f n) = sum p (fun n ↦ f' n) := by
  simp (config := {createSubgoals := true}) only [hf]
  -- desired output: 1 goal: ⊢ ∀ (n : Nat), p n → p' n 
  -- or alternatively: 1 goal: n : Nat, h† : p n ⊢ p' n
```

### Clarifications

* The discharger is applied first when using conditional rewrites, and only when the discharger fails a new goal is created.
* It should be clearly documented that this option can create unprovable goals even if the original goal was provable, which `simp` otherwise doesn't do.
* I expect that we don't want a failure of type-class inference to lead to new goals, but only propositional(?) (usually explicit) arguments that do not occur in the goal.

### Usage notes

* This is a feature that I've heard requested multiple times, and I also miss it myself. 
* It is especially useful when you are working with nested sums or integrals, since it's painful to rewrite lemmas manually
* It can also be very useful while writing a proof, so that you can see which hypotheses are needed for `simp` to successfully rewrite with a simp-lemma.

### Community Feedback

[Zulip discussion thread](https://leanprover.zulipchat.com/#narrow/channel/270676-lean4/topic/simp.20should.20be.20able.20to.20create.20new.20goals)

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others benefit from the changes in this proposal being added, please ask them to add :+1: to it.

**Comment by MichaelStollBayreuth** [comment]
With the new config syntax, one could then write `simp +createSubgoals`, which is much less typing overhead.

Is the idea that `simp` will be recursively applied to the subgoals, or would this be a separate option?

**Comment by nomeata** [comment]
I understand the motivation, but also see issues with this. For example the equational theorems for a function defined with overlapping patterns has side conditions that ensure you don't apply the wrong rewrite too eagerly. This would probably lead to unwanted, possibly unsolvable side goals with this.

This makes me wonder if there needs to be a way to more clearly indicate which hypnoses should be left around, and which really needs to be discharged for simp to apply the lemma.

Or maybe this extra complexity is better kept separate, e.g. in `aesop`?

**Comment by fpvandoorn** [comment]
> Is the idea that `simp` will be recursively applied to the subgoals, or would this be a separate option?

I was thinking not applying `simp` to the subgoals. My idea is that the discharger is applied to hypotheses first, and if it fails then (with the new option turned on) the hypothesis turns into a subgoal.


> I understand the motivation, but also see issues with this. For example the equational theorems for a function defined with overlapping patterns has side conditions that ensure you don't apply the wrong rewrite too eagerly. This would probably lead to unwanted, possibly unsolvable side goals with this.

I did not consider equational theorems with overlapping patterns. In that case it might indeed be hard to specify *which* equational theorem to apply. I don't know a good solution for this on the top of my head. Note that in Mathlib overlapping patterns happen very very rarely, so I hope this is not a blocker for this feature.

> This makes me wonder if there needs to be a way to more clearly indicate which hypnoses should be left around, and which really needs to be discharged for simp to apply the lemma.

That could be an option...

> Or maybe this extra complexity is better kept separate, e.g. in `aesop`?

I disagree. One thing I want to use this for is a "`rw` that can rewrite under binders". Now rewriting under binders seems to be inherently incompatible with the design of `rw`, but `simp only [h]` (`simp_rw [h]`) comes close to a tactic that can rewrite under binders. A major pain point is that this doesn't work with conditional equalities, so when formalizing, I spend a significant amount of time figuring out what side conditions are needed to make `simp_rw [h h']` or `simp only [h, h']` succeed. I don't think that I need to resort to `aesop` to be able to conditionally rewrite under binders. 

Lastly, I expect that `aesop` or any other tactic that would do this would have to reimplement/copy-paste a significant portion of `simp` to be able to do this (or use private internals of `simp`). I don't think the Core dev team would be happy with any of those options.

**Comment by nomeata** [comment]
Maybe the right framing for this feature is less that we want `simp` per se to support such a feature, but really this is a missing feature of `simp_rw`, and because that happens to use `simp` inside, the simp internals should become more flexible.

I wonder if we need the ability to have the `simp`-tactic-visible `Simp.Config` be a subset of the fields accepted by `Simp.Config`, should we agree that this is a useful API feature, but not something we want to expose and support for `simp`. This would be straight-forward to implement; `dsimp` already has this “subset of options exposed” syntax.

An alternative to consider would be instead of making `simp` more complex is to expose from the simplifier the building blocks that allow rewriting under binders (i.e. the congruence lemmas), and use that to implement `simp_rw`. (Unlikely the right course of action here, but an option worth keeping in mind.)

**Comment by fpvandoorn** [comment]
I expect that this option is not likely to be turned on without `only`, and I'm perfectly happy to disallow that.

**Comment by nomeata** [comment]
We have discussed this in the triage meeting, and concluded that making `simp` more complicated is the wrong step here.

Instead the goal should be that if you want to rewrite with a lemma, you can do that with `rw`, and it will just work (or give a good error message). If we had that, you wouldn't need this RFC’ed feature, right?

Of course such a change is not trivial, but we want it eventually, and thus don’t want to make changes that we wouldn’t want then anymore.

Nevertheless, as always, thanks for your contributions!

**Comment by fpvandoorn** [comment]
> Instead the goal should be that if you want to rewrite with a lemma, you can do that with `rw`, and it will just work (or give a good error message). If we had that, you wouldn't need this RFC’ed feature, right?

I'm confused about what you're saying here. Are you suggesting that we should enhance `rw` so that it can rewrite under binders? And be able to use user-defined congruence lemmas? It seems that this does not fit in the design of `rw` at all (or did you mean `simp_rw`?)

**Comment by nomeata** [comment]
Yes, that's what I am saying indeed! A unsuspecting user shouldn't need to know about the inner workings of `rw` or `simp` to be able to rewrite, should they?

This may share functionality with `simp` (and `conv`!), and likely some factoring out common functionality will need to happen.

(Now I wonder if a combination of `conv`'s `pattern`, with the pattern taken from the rule's lhs, and `rw` would work rather nicely for this usecase?)

**Comment by fpvandoorn** [comment]
Ok, awesome!

That would also be my desired ultimate outcome, a tactic that can basically do everything that `rw` or `simp_rw` can currently do, combined: both conditional rewrites and rewrite under binders.

Do you want me to make a new RFC for that? Or is it already on some internal roadmap?

**Comment by nomeata** [comment]
An rfc, just to have an issue number, might be helpful. I don't think it's near the top of anyone's roadmap, I fear.

**Comment by fpvandoorn** [comment]
Ok, I created #6033 with a (very short) new RFC.

One more data point about this RFC: 
I discussed a potential tactic with Arend Mellendijk today where we want to write a tactic that can pull binary operations (`+`, `*`, ...) out of big operators (`∑`, `∫`, ...). Some of the lemmas to do this have side conditions (for example the integrability conditions of [MeasureTheory.integral_add](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Integral/Bochner.html#MeasureTheory.integral_add)), and it would be nice to create new goals for such side conditions (if they cannot be discharged automatically). The most convenient way to implement a tactic like that is to have a custom simp-set with 10-100 lemmas and we use `simp` to eagerly apply these lemmas.

So I personally think this RFC has value beyond just the `rw` tactic that can rewrite within binders.
